src/main/java/io/temporal/internal/statemachines/SideEffectStateMachine.java

```diff
@@ -0,0 +1,164 @@
+
+  private Optional<Payloads> result;
+
+  public static final StateMachineDefinition<State, ExplicitEvent, SideEffectStateMachine>
+      STATE_MACHINE_DEFINITION =
+          StateMachineDefinition.<State, ExplicitEvent, SideEffectStateMachine>newInstance(
+                  "SideEffect", State.CREATED, State.MARKER_COMMAND_RECORDED)
+              .add(
+                  State.CREATED,
+                  ExplicitEvent.SCHEDULE,
+                  new State[] {
+                    State.MARKER_COMMAND_CREATED, State.MARKER_COMMAND_CREATED_REPLAYING
+                  },
+                  SideEffectStateMachine::createMarkerCommand)
+              .add(
+                  State.MARKER_COMMAND_CREATED_REPLAYING,
+                  CommandType.COMMAND_TYPE_RECORD_MARKER,
+                  State.RESULT_NOTIFIED_REPLAYING)
+              .add(
+                  State.MARKER_COMMAND_CREATED,
+                  CommandType.COMMAND_TYPE_RECORD_MARKER,
+                  State.RESULT_NOTIFIED,
+                  SideEffectStateMachine::markerResultFromFunc)
+              .add(
+                  State.RESULT_NOTIFIED,
+                  EventType.EVENT_TYPE_MARKER_RECORDED,
+                  State.MARKER_COMMAND_RECORDED)
+              .add(
+                  State.RESULT_NOTIFIED_REPLAYING,
+                  EventType.EVENT_TYPE_MARKER_RECORDED,
+                  State.MARKER_COMMAND_RECORDED,
+                  SideEffectStateMachine::markerResultFromEvent);
+
+  /**
+   * Creates new SideEffect Marker
+   *
+   * @param func used to produce side effect value. null if replaying.
+   * @param callback returns side effect value or failure
+   * @param commandSink callback to send commands to
+   */
+  public static void newInstance(
+      Functions.Func<Boolean> replaying,
+      Functions.Func<Optional<Payloads>> func,
+      Functions.Proc1<Optional<Payloads>> callback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    new SideEffectStateMachine(replaying, func, callback, commandSink, stateMachineSink);
+  }
+
+  private SideEffectStateMachine(
+      Functions.Func<Boolean> replaying,
+      Functions.Func<Optional<Payloads>> func,
+      Functions.Proc1<Optional<Payloads>> callback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    super(STATE_MACHINE_DEFINITION, commandSink, stateMachineSink);
+    this.replaying = replaying;
+    this.func = func;
+    this.callback = callback;
+    explicitEvent(ExplicitEvent.SCHEDULE);
+  }
+
+  private State createMarkerCommand() {
+    RecordMarkerCommandAttributes markerAttributes;
+    if (replaying.apply()) {
+      markerAttributes = RecordMarkerCommandAttributes.getDefaultInstance();
+      return State.MARKER_COMMAND_CREATED_REPLAYING;
+    } else {
+      // executing first time
+      result = func.apply();
+      if (result == null) {
```

Comment:
does this mean a side effect cannot return a null value?

Reply:
yes